File: LeetCode/[M] - 216. Combination Sum III/solution.py
from typing import List


class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        res = []

        def backtracking(i, seq):
            if len(seq) == k:
                if sum(seq) == n:
                    res.append(seq)
                return
            if sum(seq) > n:
                return

            for x in range(i, 10):
                backtracking(x + 1, seq + [x])

        backtracking(1, [])
        return res

File: LeetCode/[M] - 216. Combination Sum III/solution.go
package main

func combinationSum3(k int, n int) [][]int {
	var res [][]int
	var fn func(int, []int)

	fn = func(x int, seq []int) {
		if len(seq) == k {
			if sliceSum(seq) == n {
				tmp := make([]int, len(seq))
				copy(tmp, seq)
				res = append(res, tmp)
			}
			return
		}
		if sliceSum(seq) > n {
			return
		}

		for i := x; i < 10; i++ {
			seq = append(seq, i)
			fn(i+1, seq)
			seq = seq[:len(seq)-1]
		}
	}

	fn(1, []int(nil))

	return res
}

func sliceSum(slice []int) int {
	sum := 0
	for _, num := range slice {
		sum += num
	}
	return sum
}
